apprating/src/main/java/com/mindera/skeletoid/apprating/controller/AppRatingController.kt

```diff
@@ -0,0 +1,124 @@
+import java.util.Date
+
+/**
+ * Controller responsible for handling the stored data and the dialog itself
+ */
+class AppRatingController {
+
+    private var countsPerTimeInterval: Pair<Int, Long>? = null
+    private var promptTimeInterval: Long? = null
+
+    /**
+     * Sets up the conditions to prompt the dialog.
+     *
+     * @param countsPerTimeInterval Pair<Int, Long> values with the maximum number of times the dialog can be prompt per time range (in days)
+     * @param promptTimeInterval Time distance between prompts (in days)
+     */
+    fun setupConditions(countsPerTimeInterval: Pair<Int, Long>, promptTimeInterval: Long) {
+        this.countsPerTimeInterval = countsPerTimeInterval
+        this.promptTimeInterval = promptTimeInterval
+    }
+
+    /**
+     * Checks if the data in the store matches the conditions and updates the values in the store.
+     *
+     * @param context Context
+     * @param dialogResultCallback Callback to handle default dialog response
+     * @return true if the conditions were met, and the dialog can be prompted.
+     */
+    fun promptDialog(context: Context, dialogResultCallback: AppRatingDialogResponseCallback? = null): Boolean {
+        val store = AppRatingStore(context)
+        return shouldPromptDialog(store).also {
+            if (it) {
+                dialogResultCallback?.let {
+                    TODO("IMPLEMENT DEFAULT DIALOG")
+                }
+                store.promptedCount++
+                store.lastTimePrompted = DateUtils.formatDate(Date())
+            }
+        }
+    }
+
+    /**
+     * Handles the response of the dialog by updating the values in the store.
+     * If response is RATE_LATER it should schedule a one time work that starts after a delay equal to the promptTimeInterval (in days).
+     * If response is RATE_NOW or NEVER_RATE it simply updates the store.
+     *
+     * @param context Context
+     * @param response Selected response to the dialog
+     */
+    fun handleDialogResponse(context: Context, response: AppRatingDialogResponse) {
+        val store = AppRatingStore(context)
+        when (response) {
+            AppRatingDialogResponse.RATE_NOW, AppRatingDialogResponse.NEVER_RATE -> {
+                store.alreadyRated = true
+            }
+            AppRatingDialogResponse.RATE_LATER -> {
+                promptTimeInterval?.let { AppRatingJobInitializer.schedule(it) }
+            }
+        }
+    }
+
+    /**
+     * Checks if the values in the store matches the conditions to prompt the dialog.
+     *
+     * @param store Store that has the values related with the app rating conditions
+     * @return true if the conditions were met
+     */
+    private fun shouldPromptDialog(store: AppRatingStore): Boolean {
+        return if (store.initialPromptDate.isEmpty()) {
+            store.initialPromptDate = DateUtils.formatDate(Date())
+            store.promptedCount = 0
+            store.alreadyRated = false
```

Comment:
Functions with collateral effects doesn't seem a very good idea. Can we do this in a different way?

Reply:
What do you mean?

Reply:
I mean that functions should only do what they are suppose to do and this one checks if should prompt dialog and changes stuff on the store. These are the collateral effects we should avoid.

Reply:
🤔 Changed the responsability to the `promptDialog` fun. Wdyt?

Reply:
The promptDialog also returns a Boolean so it seems like collateral effects as well 🙈